auth_totp/controllers/main.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 LasLabs Inc.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from datetime import datetime, timedelta
+import json
+from werkzeug.contrib.securecookie import SecureCookie
+from openerp import http, registry, SUPERUSER_ID
+from openerp.api import Environment
+from openerp.http import Response, request
+from openerp.addons.web.controllers.main import Home
+from ..exceptions import MfaTokenInvalidError, MfaTokenExpiredError
+
+
+class JsonSecureCookie(SecureCookie):
+    serialization_method = json
+
+
+class AuthTotp(Home):
+    @http.route()
+    def web_login(self, *args, **kwargs):
+        response = super(AuthTotp, self).web_login(*args, **kwargs)
+
+        if not request.params.get('login_success'):
+            return response
+
+        user = request.env['res.users'].sudo().browse(request.uid)
+        if not user.mfa_enabled:
+            return response
+
+        cookie_key = 'trusted_devices_%s' % user.id
+        device_cookie = request.httprequest.cookies.get(cookie_key)
+        if device_cookie:
+            config_model = request.env['ir.config_parameter']
+            secret = config_model.sudo().get_param('database.secret')
```

Comment:
Nitpicky, but let's centralize our sudos at the top and add a comment as to why we are doing this. 

```
SudoConfig = request.env['ir.config_parameter'].sudo()
SudoUser = request.env['res.users'].sudo()
```

Sudo is a dangerous beast. We should always make this stand out as much as possible in the code, as well as provide context as to the reasoning. Otherwise, a questioning of the sudo will become a yearly occurrence in migrations.